src/serialization/tuple.h

```diff
@@ -51,119 +51,71 @@ namespace serialization
   }
 }
 
-template <template <bool> class Archive, class E0, class E1, class E2>
-inline bool do_serialize(Archive<false>& ar, std::tuple<E0,E1,E2>& p)
+template <size_t I, bool BackwardsCompat, bool W, template <bool> class Archive, typename... Ts>
+bool do_serialize_tuple_nth(Archive<W>& ar, std::tuple<Ts...>& v)
 {
-  size_t cnt;
-  ar.begin_array(cnt);
-  if (!ar.good())
-    return false;
-  if (cnt != 3)
-    return false;
+  static constexpr const size_t tuple_size = std::tuple_size<std::tuple<Ts...>>();
+  static_assert(I <= tuple_size, "bad call");
 
-  if (!::serialization::detail::serialize_tuple_element(ar, std::get<0>(p)))
-    return false;
-  if (!ar.good())
-    return false;
-  ar.delimit_array();
-  if (!::serialization::detail::serialize_tuple_element(ar, std::get<1>(p)))
-    return false;
-  if (!ar.good())
-    return false;
-  ar.delimit_array();
-  if (!::serialization::detail::serialize_tuple_element(ar, std::get<2>(p)))
-    return false;
-  if (!ar.good())
-    return false;
+  if constexpr (I == 0)
+  {
+    // If BackwardsCompat is true, we serialize the size of 3-tuples and 4-tuples
+    if constexpr (BackwardsCompat && (tuple_size == 3 || tuple_size == 4))
+    {
+      size_t cnt = tuple_size;
+      ar.begin_array(cnt);
+      if (cnt != tuple_size)
+        return false;
+    }
+    else
+    {
+      ar.begin_array();
```

Comment:
I never paid attention to this serializer before, but isn't this broken for JSON? You can't have an array because each element of the tuple could be different type. So the outter type must be an object? Is the compatibility mode always using integers (or at least something all the same)?

I _think_ it works for the binary archive, but I haven't looked too closely. I'm assuming the tests verify that the heterogenous arrays do in fact work.

Another nitpick - how are new fields added to a tuple? There's no version field or anything to change the size of the tuple later.

Reply:
> I never paid attention to this serializer before, but isn't this broken for JSON? You can't have an array because each element of the tuple could be different type. So the outter type must be an object?

From [RFC 8259, Section 5](https://www.rfc-editor.org/rfc/rfc8259#section-5): "There is no requirement that the values in an array be of the same type."

> Is the compatibility mode always using integers (or at least something all the same)?

The compatibility mode serializes a 3 or 4 varint value before the array in a binary archive for 3-tuples and 4-tuples, respectively, since that is how the old code was written. I chose not to do this with tuples of other sizes, or if `BackwardsCompat=false` since that size value is redundant; we know the size of a tuple at compile time. 

 > Another nitpick - how are new fields added to a tuple? There's no version field or anything to change the size of the tuple later.

They aren't. An N-tuple is an N-tuple. IMO, if you want to serialize an object that might be extended in the future, you should just create that as a struct then use an custom serializer (with a version field).

Reply:
> From RFC 8259, Section 5: "There is no requirement that the values in an array be of the same type."

Egg on face for not reading the spec.

Reply:
> They aren't. An N-tuple is an N-tuple. IMO, if you want to serialize an object that might be extended in the future, you should just create that as a struct then use an custom serializer (with a version field).

Normally I would say every struct/tuple should have a version or length field "just-in-case", but given the use case reducing things to a minimum makes sense. Hopefully this works out with future revisions.